### PR TITLE
Added X-ServiceFabric handling

### LINKED ISSUE #2: [Proposal] Create a FabricReverseProxyIntegration middleware
I am basically using such a middleware as below on every service. This might make a useful addition to address [Special handling for port sharing services](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverseproxy#special-handling-for-port-sharing-services).
```
        public static IApplicationBuilder UseFabricReverseProxyIntegration(this IApplicationBuilder app) => app.Use(async (ctx, next) =>
        {
            await next();
            if (ctx.Response.StatusCode == 404)
                ctx.Response.Headers.Add("X-ServiceFabric", "ResourceNotFound");
        });
    }
```

If this proposal is accepted, I will be happy to create a proper pull request with a proper middleware, etc.

In future this middleware might also handle X-Forwarded headers, but this requires some cleanup on the Reverse Proxy side first: Azure/service-fabric-issues#113 (it might be handled just by the ASP.NET Core Basic proxy middleware). My intention would be to create a similar middleware to the IISIntegration.

### PR DESCRIPTION
**Pull request by mkosieradzki**
This resolves #2 

I have decided to add this to the existing Service Fabric integration middleware to make it as transparent to the user as possible.

There might be a possibility it is not a desired behavior for some corner cases. For example if someone wants to heavily use prefix sharing. We can provide opt-out option in that case.

Currently developer can create his/her own middleware after ServiceFabricIntegration middleware and strip X-ServiceFabric header.

### CODE REVIEW

File: src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricMiddleware.cs [hunk 1]
```diff
@@ -85,6 +85,16 @@ public ServiceFabricMiddleware(RequestDelegate next, string urlSuffix)
                     return;
                 }
 
+                context.Response.OnStarting(() =>
```
Comment (amanbha) [inline src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricMiddleware.cs]: Please make this pull request into develop branch.
Expose an option for this in ServiceFabricIntegrationOptions (its flag), plumb it to the middleware and modify the response to add this header when the option is specified in ServiceFabricIntegrationOptions.

**Comment by mkosieradzki** [comment]
@amanbha Would you prefer:

1. New ServiceFabricIntegrationOptions class inheriting from IOptions - as it is typical for middleware pattern (you can use Configure<ServiceFabricIntegrationOptions>() ) to reconfigure and rename ServiceFabricIntegrationOptions to something like ServiceUrlSetting and make it a property of this options class.
2. Add additional flag (I don't like this idea for two reasons): flags are not very practical (especially when number of options grow and some of them will be non-flag), and it's against the standard ASP.NET Core convention.

Also we are early with the ASP.NET Core adoption so it's the last chance to create breaking changes.

If you agree for the first option: I will create a second proper PR to update documentation: https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore as well.

**Comment by amanbha** [comment]
#1 is breaking for other customers. Lets use option 2 or create another Middleware for this as originally proposed so that users know that they are opting in.

**Comment by vturecek** [comment]
Thanks @mkosieradzki !

Ideally each middleware should do one job (the name "ServiceFabricMiddleware" is obviously too broad and we might change it in the future to be more specific), so this really should be separate middleware rather than piggyback on the unique URL middleware, with the option to have it added for you through UseServiceFabricIntegration by setting a flag. That's what happens when you use the ServiceFabricIntegrationOptions.UseUniqueUrl flag.

The idea with UseServiceFabricIntegration is that it adds all relevant SF middleware based on the flags that are passed in, but users always have the option of adding each middleware independently if they want.

So let's do two things:
 1) separate middleware
 2) add a flag to ServiceFabricIntegrationOptions that adds this middleware through UseServiceFabricIntegration.

IOptions is something we can look at a later time. There's a potential opportunity for breaking changes when we add support for asp.net core 2.0, which itself introduces breaking changes, so we can consider it then.

**Comment by mkosieradzki** [comment]
@vturecek 
IMO it's the same responsibility: it handles special HTTP Codes for Rerverse Proxy integration. And it's basically all related to URL handling. It's like declaring: "I am going to handle this URI prefix" or "Please generate uri for me".

It's also technically the same thing - for example for UseUniqueUrl the header should be always added (I can't imagine any scenario when one would want to opt-out from X-ServiceFabric header then. The same applies to 410 handling (which is already implemented in this middleware).

The None option is more tricky:
One could want to do portsharing then.


To sum up:
We want to add X-ServiceFabric always for UseUniqueUrl 

We want to invent something for None...

**Comment by mkosieradzki** [comment]
@vturecek 
Maybe this way, as you suggested:
1. Separate middleware
2. (But) Always add this middleware when UseUniqueUrl is used. Without inventing a new flag.

If one wants to use to use None, they can use this middleware directly...

But the more I think about it the more I am convinced this should work only for UseUniqueUrl...

**Comment by vturecek** [comment]
We're really talking about two separate things:

 1. UseUniqueUrl is used to help prevent clients from accidentally connecting to the wrong service. This actually has nothing to do with the Reverse Proxy. Users will want to use this option for back-end services whether they are using the Reverse Proxy to connect to those services, or connecting to them directly.

 2. X-ServiceFabric is a header that specifically and only for the Reverse Proxy to help it understand how to handle 404 responses. 

Users may want to set UseUniqueUrl but not have the X-ServiceFabric header if they do not intend to use the Reverse Proxy (or have it disabled entirely, which is an option too).

A new flag would allow users to enable this middleware only if they are in fact using the reverse proxy.

**Comment by mkosieradzki** [comment]
@vturecek I agree! I will create a separate middleware then. Thanks!

**Comment by mkosieradzki** [comment]
@vturecek I have made adjustments accordingly. I agree that middleware separation was a good idea.

I hope you don't mind I have tidied up some minor namespaces sorting etc, according to standards from ASP.NET Core repo.

File: src/Microsoft.ServiceFabric.AspNetCore/AspNetCoreCommunicationListener.cs [hunk 2]
```diff
@@ -2,19 +2,19 @@
 // Copyright (c) Microsoft Corporation.  All rights reserved.
 // Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
 // ------------------------------------------------------------
+using System;
```
Comment (amanbha) [inline src/Microsoft.ServiceFabric.AspNetCore/AspNetCoreCommunicationListener.cs]: Could you please move these usings inside the namespace below, this is the convention followed in Service fabric repositories in external as well as internal.

File: src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricSetupFilter.cs [hunk 3]
```diff
@@ -2,26 +2,32 @@
...
-    using Microsoft.AspNetCore.Builder;
-    using Microsoft.AspNetCore.Hosting;
-
     internal class ServiceFabricSetupFilter : IStartupFilter
     {
         private readonly string urlSuffix;
+        private readonly bool enableReverseProxyIntegration;
 
-        internal ServiceFabricSetupFilter(string urlSuffix)
+        internal ServiceFabricSetupFilter(string urlSuffix, bool enableReverseProxyIntegration)
```
Comment (amanbha) [inline src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricSetupFilter.cs]: Use ServiceFabricIntegrationOptions as the argument here and then use it to determine which middleware needs to be enabled, this will make future additions of middlewares if needed easier.

File: src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricSetupFilter.cs [hunk 4]
```diff
@@ -2,26 +2,32 @@
...
         {
             this.urlSuffix = urlSuffix;
+            this.enableReverseProxyIntegration = enableReverseProxyIntegration;
         }
 
         public Action<IApplicationBuilder> Configure(Action<IApplicationBuilder> next)
         {
             return app =>
             {
                 app.UseServiceFabricMiddleware(this.urlSuffix);
```
Comment (amanbha) [inline src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricSetupFilter.cs]: Enable these middlewares based on the flags in ServiceFabricIntegrationOptions passed as argument.

File: src/Microsoft.ServiceFabric.AspNetCore/WebHostBuilderServiceFabricExtension.cs [hunk 5]
```diff
@@ -50,7 +49,7 @@ public static IWebHostBuilder UseServiceFabricIntegration(this IWebHostBuilder h
             hostBuilder.ConfigureServices(services =>
             {
                 // Configure MiddleWare
-                services.AddSingleton<IStartupFilter>(new ServiceFabricSetupFilter(listener.UrlSuffix));
+                services.AddSingleton<IStartupFilter>(new ServiceFabricSetupFilter(listener.UrlSuffix, options.HasFlag(ServiceFabricIntegrationOptions.UseReverseProxyIntegration)));
```
Comment (amanbha) [inline src/Microsoft.ServiceFabric.AspNetCore/WebHostBuilderServiceFabricExtension.cs]: pass the options directly.

File: src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricReverseProxyIntegrationMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.Response.OnStarting(() =>
+            {
+                if (context.Response.StatusCode == StatusCodes.Status404NotFound)
+                {
+                    context.Response.Headers["X-ServiceFabric"] = "ResourceNotFound";
```
Comment (amanbha) [inline src/Microsoft.ServiceFabric.AspNetCore/ServiceFabricReverseProxyIntegrationMiddleware.cs]: make these strings as const

File: src/Microsoft.ServiceFabric.AspNetCore/AspNetCoreCommunicationListener.cs (hunk 2 above)
Comment (mkosieradzki) [inline src/Microsoft.ServiceFabric.AspNetCore/AspNetCoreCommunicationListener.cs]: Sorry for that. I have fixed it and only sorted usings now.

**Comment by amanbha** [comment]
Thanks @mkosieradzki . I will be adding unit tests project to the repo soon, feel free to add unit tests for this middleware to the unittests project if you have some.

**Comment by Mardoxx** [comment]
This is great, good work!

**Comment by mariwing** [comment]
Sorry if this is covered elsewhere, but I was unable to find the answear. When will this be included in a released nuget package? Thanks!

**Comment by Mardoxx** [comment]
Also wondering this, becoming a bit tedious to add X-ServiceFabric on my 404 responses! 😝

**Comment by Mardoxx** [comment]
@mariwing Seems to be in 2.7.198
